### PR TITLE
fix ensureWork cyclomatic complexity

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
fix cyclomatic complexity

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
The test is failing. Just echo the log here. And let's trigger again.
```
[BasicPropagation] basic propagation testing Deployment propagation testing 
  deployment propagation testing
  /home/runner/work/karmada/karmada/test/e2e/propagationpolicy_test.go:50
STEP: Creating PropagationPolicy(karmadatest-pvw/deploy-mv5)
STEP: Creating Deployment(karmadatest-pvw/deploy-mv5)
STEP: Waiting for deployment(karmadatest-pvw/deploy-mv5) synced on member clusters
I1222 05:40:06.495114   44802 deployment.go:43] Waiting for deployment(karmadatest-pvw/deploy-mv5) synced on cluster(member1)
I1222 05:40:11.522705   44802 deployment.go:43] Waiting for deployment(karmadatest-pvw/deploy-mv5) synced on cluster(member2)
I1222 05:40:11.552753   44802 deployment.go:43] Waiting for deployment(karmadatest-pvw/deploy-mv5) synced on cluster(member3)
STEP: Updating Deployment(karmadatest-pvw/deploy-mv5)'s replicas to 6
STEP: Waiting for deployment(karmadatest-pvw/deploy-mv5) synced on member clusters
I1222 05:40:11.808384   44802 deployment.go:43] Waiting for deployment(karmadatest-pvw/deploy-mv5) synced on cluster(member1)
I1222 05:40:16.830628   44802 deployment.go:43] Waiting for deployment(karmadatest-pvw/deploy-mv5) synced on cluster(member2)
I1222 05:40:16.854603   44802 deployment.go:43] Waiting for deployment(karmadatest-pvw/deploy-mv5) synced on cluster(member3)
STEP: Removing Deployment(karmadatest-pvw/deploy-mv5)
STEP: Check if deployment(karmadatest-pvw/deploy-mv5) diappeare on member clusters
I1222 05:40:16.975819   44802 deployment.go:67] Waiting for deployment disappear on cluster(member1)
I1222 05:40:22.020430   44802 deployment.go:67] Waiting for deployment disappear on cluster(member2)
I1222 05:40:22.026848   44802 deployment.go:67] Waiting for deployment disappear on cluster(member3)
STEP: Removing PropagationPolicy(karmadatest-pvw/deploy-mv5)

• [SLOW TEST:15.819 seconds]
[BasicPropagation] basic propagation testing
/home/runner/work/karmada/karmada/test/e2e/propagationpolicy_test.go:30
  Deployment propagation testing
  /home/runner/work/karmada/karmada/test/e2e/propagationpolicy_test.go:31
    deployment propagation testing
    /home/runner/work/karmada/karmada/test/e2e/propagationpolicy_test.go:50
------------------------------
[BasicPropagation] basic propagation testing Service propagation testing 
  service propagation testing
  /home/runner/work/karmada/karmada/test/e2e/propagationpolicy_test.go:89
STEP: Creating PropagationPolicy(karmadatest-pvw/service-8xz)
STEP: Creating Service(karmadatest-pvw/service-8xz)
STEP: Waiting for service(karmadatest-pvw/service-8xz) synced on member clusters
I1222 05:40:22.227215   44802 service.go:39] Waiting for service(karmadatest-pvw/service-8xz) synced on cluster(member1)

• Failure [300.143 seconds]
[BasicPropagation] basic propagation testing
/home/runner/work/karmada/karmada/test/e2e/propagationpolicy_test.go:30
  Service propagation testing
  /home/runner/work/karmada/karmada/test/e2e/propagationpolicy_test.go:70
    service propagation testing [It]
    /home/runner/work/karmada/karmada/test/e2e/propagationpolicy_test.go:89

    Timed out after 300.001s.
    Expected
        <bool>: false
    to equal
        <bool>: true

    /home/runner/work/karmada/karmada/test/e2e/framework/service.go:46
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS

Summarizing 1 Failure:

[Fail] [BasicPropagation] basic propagation testing Service propagation testing [It] service propagation testing 
/home/runner/work/karmada/karmada/test/e2e/framework/service.go:46

Ran 2 of 35 Specs in 316.948 seconds
FAIL! -- 1 Passed | 1 Failed | 0 Pending | 33 Skipped
--- FAIL: TestE2E (317.09s)
```

**Comment by RainbowMango** [comment]
This is the 2th failing log. I think there is something wrong with the patch.
```
• [SLOW TEST:10.758 seconds]
propagation with label and group constraints testing
/home/runner/work/karmada/karmada/test/e2e/scheduling_test.go:36
  Deployment propagation testing
  /home/runner/work/karmada/karmada/test/e2e/scheduling_test.go:37
    deployment propagation with label and group constraints testing
    /home/runner/work/karmada/karmada/test/e2e/scheduling_test.go:70
------------------------------
propagation with label and group constraints testing CustomResourceDefinition propagation testing 
  crd with specified label and group constraints propagation testing
  /home/runner/work/karmada/karmada/test/e2e/scheduling_test.go:148
STEP: Creating ClusterPropagationPolicy(foo4xxs.example-wcm.karmada.io)
STEP: Creating crd(foo4xxs.example-wcm.karmada.io)
STEP: Get crd(foo4xxs.example-wcm.karmada.io)
STEP: collect the target clusters in cluster resource binding
crd kind is CustomResourceDefinition, name is foo4xxs.example-wcm.karmada.io
MaxGroups= 1, MinGroups= 1
target clusters in cluster resource binding are [member1]
STEP: check if the scheduled condition is true
STEP: check if crd present on right clusters
I1222 06:34:20.143405   42582 scheduling_test.go:188] Waiting for crd(foo4xxs.example-wcm.karmada.io) present on cluster(member1)

• Failure [305.393 seconds]
propagation with label and group constraints testing
/home/runner/work/karmada/karmada/test/e2e/scheduling_test.go:36
  CustomResourceDefinition propagation testing
  /home/runner/work/karmada/karmada/test/e2e/scheduling_test.go:110
    crd with specified label and group constraints propagation testing [It]
    /home/runner/work/karmada/karmada/test/e2e/scheduling_test.go:148

    Unexpected error:
        <*errors.errorString | 0xc00020e2c0>: {
            s: "timed out waiting for the condition",
        }
        timed out waiting for the condition
    occurred

    /home/runner/work/karmada/karmada/test/e2e/scheduling_test.go:202
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS

Summarizing 1 Failure:

[Fail] propagation with label and group constraints testing CustomResourceDefinition propagation testing [It] crd with specified label and group constraints propagation testing 
/home/runner/work/karmada/karmada/test/e2e/scheduling_test.go:202

Ran 4 of 35 Specs in 394.175 seconds
FAIL! -- 3 Passed | 1 Failed | 0 Pending | 31 Skipped
--- FAIL: TestE2E (394.30s)
FAIL
```

**Comment by iawia002** [comment]
Maybe we need to wait for #1149 which implements the `ReviseReplica` hook, and the original code to modify the replica of Deployment and Job has been changed to use default interpreters.

File: pkg/controllers/binding/common.go [hunk 1]
```diff
@@ -101,6 +93,7 @@ func ensureWork(
 		}
 
 		workLabel := mergeLabel(clonedWorkload, workNamespace, binding, scope)
+		hasScheduledReplica, desireReplicaInfos := getReplicaInfos(targetClusters)
```
Comment (iawia002) [inline pkg/controllers/binding/common.go:96]: Why move it into the loop, it does not need to be calculated every time

File: pkg/controllers/binding/common.go [hunk 2]
```diff
@@ -110,8 +103,8 @@ func ensureWork(
 				return err
 			}
 
-			if clonedWorkload.GetKind() == util.JobKind && jobHasCompletions {
-				// For a work queue Job that usually leaves .spec.completions unset, in that case, we skip setting this field.
+			if len(jobCompletions) > 0 {
+				// For a work queue job that usually leaves .spec.completions unset, in that case, we skip setting this field.
```
Comment (iawia002) [inline pkg/controllers/binding/common.go:107]: It's better not to change this, `Job` means the `Job` object

File: pkg/controllers/binding/common.go (hunk 1 above)
Comment (kerthcet) [reply pkg/controllers/binding/common.go:96]: oops, big mistake. fixed.

File: pkg/controllers/binding/common.go [hunk 3]
```diff
@@ -70,20 +68,15 @@ func ensureWork(
...
+		jobCompletions, err = divideReplicasByJobCompletions(workload, targetClusters)
 		if err != nil {
 			return err
 		}
-		if found {
-			jobCompletions = util.DivideReplicasByTargetCluster(targetClusters, int32(completions))
-			jobHasCompletions = true
-		}
 	}
+	hasScheduledReplica, desireReplicaInfos := getReplicaInfos(targetClusters)
```
Comment (iawia002) [inline pkg/controllers/binding/common.go:79]: Can we not adjust the position of this line of code so that the whole PR diff looks clearer?

Comment (kerthcet) [reply pkg/controllers/binding/common.go:79]: Variables closer to the calling function should be more clearly I think, wdyt?

**Comment by kerthcet** [comment]
Where are we now? Should I close this PR as I heard we are doing some refactoring work now? @RainbowMango

**Comment by RainbowMango** [comment]
Sorry I missed this PR from my radar.
I'll look at it today.

File: pkg/controllers/binding/common.go [hunk 4]
```diff
@@ -110,7 +103,7 @@ func ensureWork(
 				return err
 			}
 
-			if clonedWorkload.GetKind() == util.JobKind && jobHasCompletions {
+			if len(jobCompletions) > 0 {
 				// For a work queue Job that usually leaves .spec.completions unset, in that case, we skip setting this field.
 				// Refer to: https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs.
```
Comment (RainbowMango) [inline pkg/controllers/binding/common.go:108]:
```suggestion
			// Set allocated completions for Job only when the '.spec.completions' field not omitted from resource template.
			// For jobs running with a 'work queue' usually leaves '.spec.completions' unset, in that case we skip
			// setting this field as well.
			// Refer to: https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs.
			if len(jobCompletions) > 0 {
```

Comment (RainbowMango) [reply pkg/controllers/binding/common.go:108]: I think the comment should be added above the `if condition`, I added some details.

**Review by RainbowMango — COMMENTED** [review]
Generally looks good to me.

File: pkg/controllers/binding/common.go (hunk 4 above)
Comment (kerthcet) [reply pkg/controllers/binding/common.go]: updated and squashed the commits.

**Review by RainbowMango — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**